File: scripts/frame_yaml_persistence.py
from __future__ import annotations

import pathlib
from typing import Any

import yaml


# Canonical YAML projection for editor style names.
# This is a persistence helper derived from docs/frame-classes.md and the
# shared resolver semantics, not an independent visual authority.
STYLE_SEMANTICS: dict[str, dict[str, Any]] = {
    "default": {"level": 1, "fill": "white", "border": "solid"},
    "parent": {"level": 2, "fill": "grey", "border": "solid"},
    "section": {"level": 3, "fill": "white", "border": "solid"},
    "annotation": {"level": None, "fill": "white", "border": "none"},
    "highlight": {"level": None, "fill": "black", "border": "none"},
}

SUPPORTED_FRAME_KEYS = {
    "direction",
    "gap",
    "padding",
    "padding_top",
    "padding_right",
    "padding_bottom",
    "padding_left",
    "sizing",
    "sizing_w",
    "sizing_h",
    "fill_weight",
    "align",
    "wrap",
    "width",
    "height",
    "min_width",
    "max_width",
    "max_width_chars",
    "min_height",
    "max_height",
    "fill",
    "border",
    "level",
    "position",
    "x",
    "y",
    "children_order",
    "text",
    "style",
}

UNSUPPORTED_FRAME_KEYS = {"dx", "dy", "dw", "dh", "waypoints"}
SUPPORTED_GRID_KEYS = {"cols", "col_gap", "row_gap", "outer_margin"}
IGNORED_GRID_KEYS = {"link_to_root"}
UNSUPPORTED_GRID_KEYS = {"rows", "slack_absorption"}

_LOWER_KEYS = {"direction", "sizing", "sizing_w", "sizing_h", "fill", "border", "position"}
_INT_KEYS = {
    "gap",
    "padding",
    "padding_top",
    "padding_right",
    "padding_bottom",
    "padding_left",
    "width",
    "height",
    "min_width",
    "max_width",
    "max_width_chars",
    "min_height",
    "max_height",
    "level",
    "x",
    "y",
}


def _normalise_style_name(style_name: Any) -> str | None:
    if not isinstance(style_name, str):
        return None
    style_name = style_name.strip()
    if not style_name:
        return None
    return style_name


def _style_semantics(style_name: Any) -> dict[str, Any] | None:
    canonical = _normalise_style_name(style_name)
    if canonical is None:
        return None
    semantic = STYLE_SEMANTICS.get(canonical)
    if semantic is None:
        return None
    result = dict(semantic)
    result["style"] = canonical
    return result


def _coerce_int(value: Any, field_name: str) -> int:
    try:
        return int(value)
    except (TypeError, ValueError) as exc:  # pragma: no cover - defensive
        raise ValueError(f"{field_name} must be an integer") from exc


def _coerce_float(value: Any, field_name: str) -> float:
    try:
        return float(value)
    except (TypeError, ValueError) as exc:  # pragma: no cover - defensive
        raise ValueError(f"{field_name} must be numeric") from exc


def _yaml_align(value: Any) -> str:
    if not isinstance(value, str):
        raise ValueError("align must be a string")
    return value.strip().lower().replace("_", "-")


def _yaml_text_scalar(existing: Any, text: str) -> Any:
    if isinstance(existing, dict):
        updated = dict(existing)
        updated["text"] = text
        return updated
    return text


def _update_text_fields(frame_data: dict[str, Any], text_override: Any) -> None:
    if not isinstance(text_override, dict):
        raise ValueError("text override must be an object")
    if "heading" in text_override:
        heading = text_override["heading"]
        if heading is None or heading == "":
            frame_data.pop("heading", None)
        elif not isinstance(heading, str):
            raise ValueError("text.heading must be a string")
        else:
            frame_data["heading"] = _yaml_text_scalar(frame_data.get("heading"), heading)
    if "label" in text_override:
        label = text_override["label"]
        if not isinstance(label, list) or any(not isinstance(line, str) for line in label):
            raise ValueError("text.label must be a list of strings")
        existing = frame_data.get("label")
        existing_lines = existing if isinstance(existing, list) else [existing] if existing is not None else []
        frame_data["label"] = [
            _yaml_text_scalar(existing_lines[index] if index < len(existing_lines) else None, line)
            for index, line in enumerate(label)
        ]


def _reorder_children(frame_data: dict[str, Any], child_order: Any, frame_id: str) -> None:
    if not isinstance(child_order, list) or any(not isinstance(item, str) for item in child_order):
        raise ValueError("children_order must be a list of child ids")
    children = frame_data.get("children")
    if not isinstance(children, list):
        raise ValueError(f"{frame_id} has no children to reorder")
    child_map = {
        child.get("id"): child
        for child in children
        if isinstance(child, dict) and isinstance(child.get("id"), str)
    }
    missing = [child_id for child_id in child_order if child_id not in child_map]
    if missing:
        raise ValueError(f"{frame_id} children_order references unknown child ids: {', '.join(missing)}")
    reordered = [child_map[child_id] for child_id in child_order]
    remaining = [
        child for child in children
        if not isinstance(child, dict) or child.get("id") not in child_map or child.get("id") not in child_order
    ]
    frame_data["children"] = reordered + remaining


def _apply_grid_overrides(document: dict[str, Any], grid_overrides: Any) -> None:
    if not isinstance(grid_overrides, dict):
        raise ValueError("grid_overrides must be an object")
    grid = document.get("grid")
    if not isinstance(grid, dict):
        grid = {}
        document["grid"] = grid

    for key in UNSUPPORTED_GRID_KEYS:
        if key in grid_overrides and grid_overrides.get(key) not in (None, True):
            raise ValueError(f"{key} is not persistable in frame YAML")
    if "link_to_root" in grid_overrides and grid_overrides.get("link_to_root") not in (None, True):
        raise ValueError("link_to_root=false is not persistable in frame YAML")

    margin_keys = ["margin_top", "margin_right", "margin_bottom", "margin_left"]
    margin_values = [grid_overrides.get(key) for key in margin_keys if key in grid_overrides]
    if margin_values:
        numeric_margins = [_coerce_int(value, "grid_overrides.margin") for value in margin_values]
        if len(set(numeric_margins)) != 1:
            raise ValueError("per-side grid margins are not persistable in frame YAML; values must be uniform")
        grid["outer_margin"] = numeric_margins[0]

    for key in grid_overrides:
        if key in SUPPORTED_GRID_KEYS:
            grid[key] = _coerce_int(grid_overrides[key], f"grid_overrides.{key}")
        elif key not in IGNORED_GRID_KEYS and key not in margin_keys and key not in UNSUPPORTED_GRID_KEYS:
            raise ValueError(f"Unknown grid_overrides key: {key}")


def _apply_style_fields(frame_data: dict[str, Any], style_name: Any) -> None:
    frame_data.pop("style", None)
    semantic = _style_semantics(style_name)
    if semantic is None:
        frame_data.pop("level", None)
        frame_data.pop("fill", None)
        frame_data.pop("border", None)
        return
    if semantic["level"] is None:
        frame_data.pop("level", None)
    else:
        frame_data["level"] = semantic["level"]
    frame_data["fill"] = semantic["fill"]
    frame_data["border"] = semantic["border"]


def _apply_direct_field(frame_data: dict[str, Any], key: str, value: Any) -> None:
    if key == "align":
        frame_data[key] = _yaml_align(value)
        return
    if key == "fill_weight":
        coerced = _coerce_float(value, key)
        frame_data[key] = int(coerced) if coerced.is_integer() else coerced
        return
    if key == "wrap":
        frame_data[key] = bool(value)
        return
    if key in _LOWER_KEYS:
        if not isinstance(value, str):
            raise ValueError(f"{key} must be a string")
        frame_data[key] = value.strip().lower()
        return
    if key in _INT_KEYS:
        frame_data[key] = _coerce_int(value, key)
        return
    raise ValueError(f"Unsupported canonical field: {key}")


def _find_frame_data(frame_data: dict[str, Any], frame_id: str) -> dict[str, Any] | None:
    if frame_data.get("id") == frame_id:
        return frame_data
    children = frame_data.get("children", [])
    if not isinstance(children, list):
        return None
    for child in children:
        if not isinstance(child, dict):
            continue
        found = _find_frame_data(child, frame_id)
        if found is not None:
            return found
    return None


def _apply_frame_override(frame_data: dict[str, Any], override: Any, frame_id: str) -> None:
    if not isinstance(override, dict):
        raise ValueError(f"Override for {frame_id} must be an object")

    unsupported = sorted(key for key in override if key in UNSUPPORTED_FRAME_KEYS)
    if unsupported:
        raise ValueError(
            f"{frame_id} includes non-canonical transient keys that cannot be saved to YAML: {', '.join(unsupported)}"
        )

    style_name = override.get("style")
    if "style" in override:
        _apply_style_fields(frame_data, style_name)

    for key, value in override.items():
        if key not in SUPPORTED_FRAME_KEYS and key not in UNSUPPORTED_FRAME_KEYS:
            raise ValueError(f"Unknown override key for {frame_id}: {key}")
        if key == "style":
            continue
        if key == "children_order":
            _reorder_children(frame_data, value, frame_id)
            continue
        if key == "text":
            _update_text_fields(frame_data, value)
            continue
        if key == "sizing":
            _apply_direct_field(frame_data, "sizing_w", value)
            _apply_direct_field(frame_data, "sizing_h", value)
            frame_data.pop("sizing", None)
            continue
        if key == "padding":
            _apply_direct_field(frame_data, key, value)
            for per_side in ("padding_top", "padding_right", "padding_bottom", "padding_left"):
                frame_data.pop(per_side, None)
            continue
        if key in {"level", "fill", "border"} and "style" in override:
            # Style mapping already wrote the canonical YAML values.
            continue
        if key in {
            "direction",
            "gap",
            "padding_top",
            "padding_right",
            "padding_bottom",
            "padding_left",
            "sizing_w",
            "sizing_h",
            "fill_weight",
            "align",
            "wrap",
            "width",
            "height",
            "min_width",
            "max_width",
    "max_width_chars",
            "min_height",
            "max_height",
            "fill",
            "border",
            "level",
            "position",
            "x",
            "y",
        }:
            _apply_direct_field(frame_data, key, value)


def persist_override_payload_to_yaml(
    frame_path: pathlib.Path,
    payload: dict[str, Any],
) -> None:
    if not isinstance(payload, dict):
        raise ValueError("Expected JSON object")

    overrides = payload.get("overrides", {})
    if not isinstance(overrides, dict):
        raise ValueError("overrides must be an object")
    grid_overrides = payload.get("grid_overrides")
    has_grid_overrides = "grid_overrides" in payload and isinstance(grid_overrides, dict) and len(grid_overrides) > 0
    if not overrides and not has_grid_overrides:
        return

    document = yaml.safe_load(frame_path.read_text(encoding="utf-8"))
    if not isinstance(document, dict):
        raise ValueError(f"{frame_path}: expected top-level mapping")
    if document.get("engine") != "v3":
        raise ValueError(f"{frame_path}: not a native frame YAML (missing engine: v3)")

    root_data = document.get("root")
    if not isinstance(root_data, dict):
        raise ValueError(f"{frame_path}: root must be a mapping")

    if "grid_overrides" in payload:
        _apply_grid_overrides(document, payload.get("grid_overrides"))

    for frame_id, override in overrides.items():
        if not isinstance(frame_id, str):
            raise ValueError("override ids must be strings")
        target = _find_frame_data(root_data, frame_id)
        if target is None:
            raise ValueError(f"Unknown component id in overrides: {frame_id}")
        _apply_frame_override(target, override, frame_id)

    dumped = yaml.safe_dump(document, sort_keys=False, allow_unicode=True, width=1000)
    frame_path.write_text(dumped, encoding="utf-8")

File: apps/preview/src/frame-yaml-persistence.ts
import path from "node:path";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
const yaml = require("yaml") as {
  parse: (raw: string) => unknown;
  stringify: (value: unknown, options?: Record<string, unknown>) => string;
};

const STYLE_SEMANTICS: Record<string, { level: number | null; fill: string; border: string }> = {
  default: { level: 1, fill: "white", border: "solid" },
  parent: { level: 2, fill: "grey", border: "solid" },
  section: { level: 3, fill: "white", border: "solid" },
  annotation: { level: null, fill: "white", border: "none" },
  highlight: { level: null, fill: "black", border: "none" },
};

const SUPPORTED_FRAME_KEYS = new Set([
  "direction",
  "gap",
  "padding",
  "padding_top",
  "padding_right",
  "padding_bottom",
  "padding_left",
  "sizing",
  "sizing_w",
  "sizing_h",
  "fill_weight",
  "align",
  "wrap",
  "width",
  "height",
  "min_width",
  "max_width",
  "max_width_chars",
  "min_height",
  "max_height",
  "fill",
  "border",
  "level",
  "position",
  "x",
  "y",
  "children_order",
  "text",
  "style",
]);

const UNSUPPORTED_FRAME_KEYS = new Set(["dx", "dy", "dw", "dh", "waypoints"]);
const SUPPORTED_GRID_KEYS = new Set(["cols", "col_gap", "row_gap", "outer_margin"]);
const IGNORED_GRID_KEYS = new Set(["link_to_root"]);
const UNSUPPORTED_GRID_KEYS = new Set(["rows", "slack_absorption"]);
const LOWER_KEYS = new Set(["direction", "sizing", "sizing_w", "sizing_h", "fill", "border", "position"]);
const INT_KEYS = new Set([
  "gap",
  "padding",
  "padding_top",
  "padding_right",
  "padding_bottom",
  "padding_left",
  "width",
  "height",
  "min_width",
  "max_width",
  "max_width_chars",
  "min_height",
  "max_height",
  "level",
  "x",
  "y",
]);

export interface PersistOverridePayload {
  overrides?: Record<string, unknown>;
  removed_ids?: unknown[];
  grid_overrides?: Record<string, unknown>;
  elk_layout_overrides?: Record<string, unknown>;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function normalizeStyleName(styleName: unknown): string | null {
  if (typeof styleName !== "string") return null;
  const canonical = styleName.trim();
  return canonical.length > 0 ? canonical : null;
}

function styleSemantics(styleName: unknown): Record<string, unknown> | null {
  const canonical = normalizeStyleName(styleName);
  if (!canonical) return null;
  const semantic = STYLE_SEMANTICS[canonical];
  if (!semantic) return null;
  return { ...semantic, style: canonical };
}

function coerceInt(value: unknown, fieldName: string): number {
  const numeric = Number(value);
  if (!Number.isInteger(numeric)) {
    throw new Error(`${fieldName} must be an integer`);
  }
  return numeric;
}

function coerceFloat(value: unknown, fieldName: string): number {
  const numeric = Number(value);
  if (!Number.isFinite(numeric)) {
    throw new Error(`${fieldName} must be numeric`);
  }
  return numeric;
}

function yamlAlign(value: unknown): string {
  if (typeof value !== "string") throw new Error("align must be a string");
  return value.trim().toLowerCase().replaceAll("_", "-");
}

function yamlTextScalar(existing: unknown, text: string): unknown {
  if (isRecord(existing)) {
    return { text };
  }
  return text;
}

function updateTextFields(frameData: Record<string, unknown>, textOverride: unknown): void {
  if (!isRecord(textOverride)) {
    throw new Error("text override must be an object");
  }
  if ("heading" in textOverride) {
    const heading = textOverride.heading;
    if (heading == null || heading === "") {
      delete frameData.heading;
    } else if (typeof heading !== "string") {
      throw new Error("text.heading must be a string");
    } else {
      frameData.heading = yamlTextScalar(frameData.heading, heading);
    }
  }
  if ("label" in textOverride) {
    const label = textOverride.label;
    if (!Array.isArray(label) || label.some((line) => typeof line !== "string")) {
      throw new Error("text.label must be a list of strings");
    }
    const existing = frameData.label;
    const existingLines = Array.isArray(existing) ? existing : existing != null ? [existing] : [];
    frameData.label = label.map((line, index) =>
      yamlTextScalar(index < existingLines.length ? existingLines[index] : null, line),
    );
  }
}

function reorderChildren(frameData: Record<string, unknown>, childOrder: unknown, frameId: string): void {
  if (!Array.isArray(childOrder) || childOrder.some((item) => typeof item !== "string")) {
    throw new Error("children_order must be a list of child ids");
  }
  const children = frameData.children;
  if (!Array.isArray(children)) {
    throw new Error(`${frameId} has no children to reorder`);
  }
  const childMap = new Map<string, Record<string, unknown>>();
  for (const child of children) {
    if (isRecord(child) && typeof child.id === "string") {
      childMap.set(child.id, child);
    }
  }
  const missing = childOrder.filter((childId) => !childMap.has(childId));
  if (missing.length > 0) {
    throw new Error(`${frameId} children_order references unknown child ids: ${missing.join(", ")}`);
  }
  const reordered = childOrder.map((childId) => childMap.get(childId) as Record<string, unknown>);
  const remaining = children.filter(
    (child) => !isRecord(child) || typeof child.id !== "string" || !childOrder.includes(child.id),
  );
  frameData.children = [...reordered, ...remaining];
}

function applyGridOverrides(document: Record<string, unknown>, gridOverrides: unknown): void {
  if (!isRecord(gridOverrides)) {
    throw new Error("grid_overrides must be an object");
  }
  const grid = isRecord(document.grid) ? document.grid : {};
  document.grid = grid;

  for (const key of UNSUPPORTED_GRID_KEYS) {
    if (key in gridOverrides && gridOverrides[key] !== null && gridOverrides[key] !== true) {
      throw new Error(`${key} is not persistable in frame YAML`);
    }
  }
  if ("link_to_root" in gridOverrides && gridOverrides.link_to_root !== null && gridOverrides.link_to_root !== true) {
    throw new Error("link_to_root=false is not persistable in frame YAML");
  }

  const marginKeys = ["margin_top", "margin_right", "margin_bottom", "margin_left"] as const;
  const marginValues = marginKeys.filter((key) => key in gridOverrides).map((key) => gridOverrides[key]);
  if (marginValues.length > 0) {
    const numericMargins = marginValues.map((value) => coerceInt(value, "grid_overrides.margin"));
    if (new Set(numericMargins).size !== 1) {
      throw new Error("per-side grid margins are not persistable in frame YAML; values must be uniform");
    }
    grid.outer_margin = numericMargins[0];
  }

  for (const [key, value] of Object.entries(gridOverrides)) {
    if (SUPPORTED_GRID_KEYS.has(key)) {
      grid[key] = coerceInt(value, `grid_overrides.${key}`);
    } else if (
      !IGNORED_GRID_KEYS.has(key) &&
      !UNSUPPORTED_GRID_KEYS.has(key) &&
      !marginKeys.includes(key as (typeof marginKeys)[number])
    ) {
      throw new Error(`Unknown grid_overrides key: ${key}`);
    }
  }
}

function applyStyleFields(frameData: Record<string, unknown>, styleName: unknown): void {
  delete frameData.style;
  const semantic = styleSemantics(styleName);
  if (!semantic) {
    delete frameData.level;
    delete frameData.fill;
    delete frameData.border;
    return;
  }
  if (semantic.level == null) {
    delete frameData.level;
  } else {
    frameData.level = semantic.level;
  }
  frameData.fill = semantic.fill;
  frameData.border = semantic.border;
}

function isImplicitStructuralWrapperFrame(frameData: Record<string, unknown>): boolean {
  const children = frameData.children;
  if (!Array.isArray(children) || children.length === 0) {
    return false;
  }
  const heading = frameData.heading;
  if (typeof heading === "string" && heading.trim()) return false;
  if (isRecord(heading) && String(heading.text ?? "").trim()) return false;
  return !("level" in frameData) && !("fill" in frameData) && !("border" in frameData) && !("variant" in frameData);
}

function applyDirectField(frameData: Record<string, unknown>, key: string, value: unknown): void {
  if (key === "align") {
    frameData[key] = yamlAlign(value);
    return;
  }
  if (key === "fill_weight") {
    const coerced = coerceFloat(value, key);
    frameData[key] = Number.isInteger(coerced) ? Math.trunc(coerced) : coerced;
    return;
  }
  if (key === "wrap") {
    frameData[key] = Boolean(value);
    return;
  }
  if (LOWER_KEYS.has(key)) {
    if (typeof value !== "string") throw new Error(`${key} must be a string`);
    frameData[key] = value.trim().toLowerCase();
    return;
  }
  if (INT_KEYS.has(key)) {
    frameData[key] = coerceInt(value, key);
    return;
  }
  throw new Error(`Unsupported canonical field: ${key}`);
}

function removeFrameFromTree(frameData: Record<string, unknown>, frameId: string): boolean {
  const children = frameData.children;
  if (!Array.isArray(children)) return false;
  for (let index = 0; index < children.length; index += 1) {
    const child = children[index];
    if (!isRecord(child)) continue;
    if (child.id === frameId) {
      children.splice(index, 1);
      return true;
    }
    if (removeFrameFromTree(child, frameId)) return true;
  }
  return false;
}

function applyRemovedFrameIds(document: Record<string, unknown>, removedIds: string[]): void {
  if (removedIds.length === 0) return;
  const rootData = document.root;
  if (!isRecord(rootData)) throw new Error("root must be a mapping");
  const rootId = typeof rootData.id === "string" ? rootData.id : "";
  const removed = new Set(removedIds.filter((frameId) => typeof frameId === "string" && frameId));
  if (rootId && removed.has(rootId)) {
    throw new Error(`Cannot remove diagram root frame: ${rootId}`);
  }
  for (const frameId of [...removed].sort()) {
    if (frameId !== rootId) {
      removeFrameFromTree(rootData, frameId);
    }
  }
  if (Array.isArray(document.arrows)) {
    document.arrows = document.arrows.filter(
      (arrow) =>
        isRecord(arrow) &&
        !removed.has(String(arrow.source ?? "")) &&
        !removed.has(String(arrow.target ?? "")),
    );
  }
}

function findFrameData(frameData: Record<string, unknown>, frameId: string): Record<string, unknown> | null {
  if (frameData.id === frameId) return frameData;
  const children = frameData.children;
  if (!Array.isArray(children)) return null;
  for (const child of children) {
    if (!isRecord(child)) continue;
    const found = findFrameData(child, frameId);
    if (found) return found;
  }
  return null;
}

function applyFrameOverride(frameData: Record<string, unknown>, override: unknown, frameId: string): void {
  if (!isRecord(override)) {
    throw new Error(`Override for ${frameId} must be an object`);
  }
  const unsupported = Object.keys(override).filter((key) => UNSUPPORTED_FRAME_KEYS.has(key)).sort();
  if (unsupported.length > 0) {
    throw new Error(
      `${frameId} includes non-canonical transient keys that cannot be saved to YAML: ${unsupported.join(", ")}`,
    );
  }

  const styleName = override.style;
  if ("style" in override) {
    if (isImplicitStructuralWrapperFrame(frameData) && normalizeStyleName(styleName)) {
      // no-op: keep wrappers structural-only
    } else {
      applyStyleFields(frameData, styleName);
    }
  }

  for (const [key, value] of Object.entries(override)) {
    if (!SUPPORTED_FRAME_KEYS.has(key) && !UNSUPPORTED_FRAME_KEYS.has(key)) {
      throw new Error(`Unknown override key for ${frameId}: ${key}`);
    }
    if (key === "style") continue;
    if (key === "children_order") {
      reorderChildren(frameData, value, frameId);
      continue;
    }
    if (key === "text") {
      updateTextFields(frameData, value);
      continue;
    }
    if (key === "sizing") {
      applyDirectField(frameData, "sizing_w", value);
      applyDirectField(frameData, "sizing_h", value);
      delete frameData.sizing;
      continue;
    }
    if (key === "padding") {
      applyDirectField(frameData, key, value);
      delete frameData.padding_top;
      delete frameData.padding_right;
      delete frameData.padding_bottom;
      delete frameData.padding_left;
      continue;
    }
    if ((key === "level" || key === "fill" || key === "border") && "style" in override) {
      continue;
    }
    if (
      [
        "direction",
        "gap",
        "padding_top",
        "padding_right",
        "padding_bottom",
        "padding_left",
        "sizing_w",
        "sizing_h",
        "fill_weight",
        "align",
        "wrap",
        "width",
        "height",
        "min_width",
        "max_width",
        "max_width_chars",
        "min_height",
        "max_height",
        "fill",
        "border",
        "level",
        "position",
        "x",
        "y",
      ].includes(key)
    ) {
      applyDirectField(frameData, key, value);
    }
  }
}

function applyElkLayoutOverrides(document: Record<string, unknown>, elkOverrides: Record<string, unknown>): void {
  if (Object.keys(elkOverrides).length === 0) return;
  const meta = isRecord(document.meta) ? document.meta : {};
  document.meta = meta;
  const elk: Record<string, string> = isRecord(meta.elk)
    ? Object.fromEntries(Object.entries(meta.elk).map(([key, value]) => [String(key), String(value)]))
    : {};
  for (const [key, value] of Object.entries(elkOverrides)) {
    if (value == null || String(value) === "") {
      delete elk[String(key)];
    } else {
      elk[String(key)] = String(value);
    }
  }
  if (Object.keys(elk).length > 0) {
    meta.elk = elk;
  } else {
    delete meta.elk;
  }
}

export function verifyElkLayoutPersisted(documentText: string, expected: Record<string, unknown>): void {
  if (Object.keys(expected).length === 0) return;
  const document = yaml.parse(documentText);
  if (!isRecord(document)) throw new Error("expected top-level mapping after save");
  if (!isRecord(document.meta)) throw new Error("meta missing after ELK save");
  if (!isRecord(document.meta.elk)) throw new Error("meta.elk missing after ELK save");
  for (const [key, raw] of Object.entries(expected)) {
    const want = String(raw);
    const got = document.meta.elk[key];
    if (got == null) {
      throw new Error(`meta.elk missing key ${JSON.stringify(key)} after save`);
    }
    if (String(got) !== want) {
      throw new Error(`meta.elk[${JSON.stringify(key)}] is ${JSON.stringify(got)}, expected ${JSON.stringify(want)} after save`);
    }
  }
}

export function persistOverridePayloadToYaml(
  framePath: string,
  baselineText: string,
  payload: PersistOverridePayload,
): string {
  if (!isRecord(payload)) throw new Error("Expected JSON object");
  const overrides = isRecord(payload.overrides) ? payload.overrides : {};
  if ("overrides" in payload && !isRecord(payload.overrides)) {
    throw new Error("overrides must be an object");
  }
  const removedIds = payload.removed_ids == null ? [] : payload.removed_ids;
  if (!Array.isArray(removedIds)) {
    throw new Error("removed_ids must be an array");
  }
  const gridOverrides = payload.grid_overrides;
  const hasGridOverrides = isRecord(gridOverrides) && Object.keys(gridOverrides).length > 0;
  const elkLayoutOverrides = payload.elk_layout_overrides;
  const hasElkOverrides = isRecord(elkLayoutOverrides) && Object.keys(elkLayoutOverrides).length > 0;
  if (Object.keys(overrides).length === 0 && !hasGridOverrides && removedIds.length === 0 && !hasElkOverrides) {
    return baselineText;
  }

  const document = yaml.parse(baselineText);
  if (!isRecord(document)) throw new Error(`${framePath}: expected top-level mapping`);
  if (document.engine !== "v3") {
    throw new Error(`${framePath}: not a native frame YAML (missing engine: v3)`);
  }
  const rootData = document.root;
  if (!isRecord(rootData)) throw new Error(`${framePath}: root must be a mapping`);

  if ("grid_overrides" in payload) {
    applyGridOverrides(document, gridOverrides ?? {});
  }
  if (hasElkOverrides) {
    applyElkLayoutOverrides(document, elkLayoutOverrides);
  }
  if (removedIds.length > 0) {
    applyRemovedFrameIds(
      document,
      removedIds.filter((frameId): frameId is string => typeof frameId === "string"),
    );
  }
  for (const [frameId, override] of Object.entries(overrides)) {
    const target = findFrameData(rootData, frameId);
    if (!target) {
      throw new Error(`Unknown component id in overrides: ${frameId}`);
    }
    applyFrameOverride(target, override, frameId);
  }

  return yaml.stringify(document, {
    aliasDuplicateObjects: false,
    lineWidth: 1000,
    sortMapEntries: false,
  });
}

export function persistForceAuthoredSpecToYaml(
  framePath: string,
  payload: unknown,
): string {
  if (!isRecord(payload) || !Array.isArray(payload.nodes) || !Array.isArray(payload.links)) {
    throw new Error("Expected authored force spec JSON payload");
  }
  const simulation = payload.simulation;
  if (isRecord(simulation) && isRecord(simulation.params)) {
    throw new Error("Expected authored force spec JSON payload, not runtime snapshot state");
  }
  return yaml.stringify(payload, {
    aliasDuplicateObjects: false,
    lineWidth: 1000,
    sortMapEntries: false,
  });
}

export function fileLabelForError(filePath: string): string {
  return path.basename(filePath);
}
